File: FEM/BST.js
var Nuts = /** @class */ (function () {
    function Nuts(value) {
        this.value = value;
        this.left = null;
        this.right = null;
    }
    return Nuts;
}());
var BST = /** @class */ (function () {
    function BST() {
        this.root = null;
    }
    BST.prototype.insert = function (value) {
        var newNode = new Nuts(value);
        if (this.root === null) {
            this.root = newNode;
        }
        else {
            this.insertNuts(this.root, newNode);
        }
    };
    BST.prototype.insertNuts = function (node, newNode) {
        if (newNode.value < node.value) {
            if (node.left === null) {
                node.left = newNode;
            }
            else {
                this.insertNuts(node.left, newNode);
            }
        }
        else {
            if (node.right === null) {
                node.right = newNode;
            }
            else {
                this.insertNuts(node.right, newNode);
            }
        }
    };
    BST.prototype.search = function (value) {
        return this.searchNuts(this.root, value);
    };
    BST.prototype.searchNuts = function (node, value) {
        if (node === null) {
            return false;
        }
        if (value < node.value) {
            return this.searchNuts(node.left, value);
        }
        else if (value > node.value) {
            return this.searchNuts(node.right, value);
        }
        else {
            return true;
        }
    };
    BST.prototype.inorderTraversal = function (callback) {
        this.inorderTraversalNuts(this.root, callback);
    };
    BST.prototype.inorderTraversalNuts = function (node, callback) {
        if (node !== null) {
            this.inorderTraversalNuts(node.left, callback);
            callback(node.value);
            this.inorderTraversalNuts(node.right, callback);
        }
    };
    return BST;
}());
var bst = new BST();
bst.insert(5);
bst.insert(3);
bst.insert(7);
bst.insert(2);
bst.insert(4);
bst.insert(6);
bst.insert(8);
console.log(bst.search(3));
console.log(bst.search(9));

File: dsa-python/bst.py
class TreeNode:
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
    
    def __str__(self):
        return str(self.key)

class BST:
    def __init__(self):
        self.root = None

    def insert(self, key):
        if self.root is None:
            self.root = TreeNode(key)
        else:
            self._insert_recursive(self.root, key)

    def _insert_recursive(self, curr_node, key):
        if key < curr_node.key:
            if curr_node.left is None:
                curr_node.left = TreeNode(key)
            else:
                self._insert_recursive(curr_node.left, key)
        else:
            if curr_node.right is None:
                curr_node.right = TreeNode(key)
            else:
                self._insert_recursive(curr_node.right, key)

    def search(self, key):
        return self._search_recursive(self.root, key)
    
    def _search_recursive(self, curr_node, key):
        if curr_node is None or curr_node.key == key:
            return curr_node
        if key < curr_node.key:
            return self._search_recursive(curr_node.left, key)
        else:
            return self._search_recursive(curr_node.right, key)
        
    def delete(self, key):
        if self.root is None:
            return
        self._delete_recursive(self.root, key)

    def _delete_recursive(self, curr_node, key):
        if key < curr_node.key:
            curr_node.left = self._delete_recursive(curr_node.left, key)
        elif key > curr_node.key:
            curr_node.right = self._delete_recursive(curr_node.right, key)
        else:
            if curr_node.left is None:
                temp = curr_node.right
                curr_node = None
                return temp
            elif curr_node.right is None:
                temp = curr_node.left
                curr_node = None
                return temp
            temp = self._min_val(curr_node.right)
            curr_node.right.key = temp.key
            curr_node.right = self._delete_recursive(curr_node.right, temp.key)
    
    def _min_val(self, node):
        curr = node
        while curr.left is not None:
            curr = curr.left
        return curr

    def inorder_traversal(self):
        self._inorder_recursive(self.root)
    
    def _inorder_recursive(self, curr_node):
        if curr_node is not None:
            self._inorder_recursive(curr_node.left)
            print(curr_node)
            self._inorder_recursive(curr_node.right)

    def preorder_traversal(self):
        self._preorder_recursive(self.root)
    
    def _preorder_recursive(self, curr_node):
        if curr_node is not None:
            print(curr_node)
            self._preorder_recursive(curr_node.left)
            self._preorder_recursive(curr_node.right)

    def postorder_traversal(self):
        self._postorder_recursive(self.root)
    
    def _postorder_recursive(self, curr_node):
        if curr_node is not None:
            self._postorder_recursive(curr_node.left)
            self._postorder_recursive(curr_node.right)
            print(curr_node)
    
bst = BST()

bst.insert(5)
bst.insert(3)
bst.insert(7)
bst.insert(2)
bst.insert(4)
bst.insert(6)
bst.insert(8)

print(bst.search(3))
print(bst.search(9))

bst.delete(5)
print(bst.search(5))

bst.inorder_traversal()
bst.preorder_traversal()
bst.postorder_traversal()